kadasrouting/gui/cpbottombar.py

```diff
@@ -0,0 +1,165 @@
+        self.patrolFootprint.setToGeometry(polygon)
+
+    def _radioButtonsPatrolChanged(self):
+        self.populatePatrolLayerSelector()
+        self.comboPatrolAreaLayers.setEnabled(self.radioPatrolAreaLayer.isChecked())
+        self.btnPatrolAreaCanvas.setEnabled(self.radioPatrolAreaPolygon.isChecked())
+        self.btnPatrolAreaClear.setEnabled(self.radioPatrolAreaPolygon.isChecked())
+        if self.radioPatrolAreaPolygon.isChecked():
+            if self.patrolArea is not None:
+                self.patrolFootprint.setToGeometry(self.patrolArea)
+        else:
+            self.patrolFootprint.reset(QgsWkbTypes.PolygonGeometry)
+
+    def clearPatrol(self):
+        self.patrolArea = None
+        self.patrolFootprint.reset(QgsWkbTypes.PolygonGeometry)
+
+    def prepareValhalla(self):
+        (
+            layer,
+            points,
+            profile,
+            allAreasToAvoidWGS,
+            costingOptions,
+        ) = super().prepareValhalla()
+        if self.radioPatrolAreaPolygon.isChecked():
+            # Currently only single polygon is accepted
+            patrolArea = self.patrolArea
+            canvasCrs = self.canvas.mapSettings().destinationCrs()
+            transformer = transformToWGS(canvasCrs)
+            if patrolArea is None:
+                # if the custom polygon button is checked, but no polygon has been drawn
+                pushWarning(
+                    self.tr("Custom polygon button is checked, but no polygon is drawn")
+                )
+                return
+        elif self.radioPatrolAreaLayer.isChecked():
+            patrolLayer = self.comboPatrolAreaLayers.currentData()
+            if patrolLayer is not None:
+                layerCrs = patrolLayer.crs()
+                transformer = transformToWGS(layerCrs)
+                patrolFeatures = [f for f in patrolLayer.getFeatures()]
+                if len(patrolFeatures) != 1:
+                    pushWarning(
+                        self.tr(
+                            "The polygon layer for Patrol must contain exactly only one polygon."
+                        )
+                    )
+                    return
+                else:
+                    patrolArea = patrolFeatures[0].geometry()
+            else:
+                # If polygon layer button is checked, but no layer polygon is selected
+                pushWarning(
+                    self.tr(
+                        "Polygon layer button is checked for Patrol, but no layer polygon is selected"
+                    )
+                )
+                return
+        else:
+            # No areas to avoid
+            patrolArea = None
+            patrolAreaWGS = None
+
+        # transform to WGS84 (Valhalla's requirement)
+        if patrolArea:
+            patrolAreaJson = json.loads(patrolArea.asJson())
+            patrolAreaWGS = []
+            polygon = patrolAreaJson["coordinates"][0]
+            for point in polygon:
```

Comment:
Why do you do this point-by-point transformation? self.patrolArea is a QgsGeometry, so you can directly transform it

Reply:
I think it was from my PR: https://github.com/camptocamp/kadas-routing-plugin/pull/57/files#diff-5b0bb1792a944cff74b00cc906bb0ff6e7efa23a7f559e8c041e2b7efa952618R235

Honestly, I can't remember why I did it. I only remember that the type of geometry is a MultiPolygon, but Valhalla accepts Polygon type only. But I agree with Victor here.

Reply:
good point, I'll change it and test if there are problems.

Reply:
from what I see, if you transform it directly then you get an error `QgsGeometry is not json serializable` while if we stick with the explicit list of points we don't have to do it later. I think I will leave it like this.